### DISCUSSION TITLE
stdout / stderr logs

### DISCUSSION THREAD

**Discussion by uggyuggy** [d-Rickyy-b/certstream-server-go#43] [Q&A] [answered]
Hi,

Starting the daemon, it looks all "information" logs goes into `stderr` (and nothing to `stdout`)

```
$ ./certstream-server-go_1.7.0-rc_linux_amd64 1> 111 2> 222

 $ tail 222

2024/08/19 15:08:37 server.go:287: Starting webserver on 127.0.0.1:8080
2024/08/19 15:08:37 ct-watcher.go:90: Checking for new ct logs...
...
```

Is it expected behavior to log this into `stderr` instead of  `stdout` ?

**Comment by d-Rickyy-b** [comment, marked as answer]
Go uses stderr as output by default for their basic logging library ([src/log/log.go](https://github.com/golang/go/blob/2e0a6f855b9af3441347b1b430e5a5e6f82a37fa/src/log/log.go#L89-L90)). 

This is in compliance with the POSIX standard, which defines the three streams (stdin, stdout, stderr) as follows:
>standard input (for reading conventional input), standard output (for writing conventional output), and standard error (for writing diagnostic output).

(see also https://pubs.opengroup.org/onlinepubs/9699919799/functions/stderr.html)

This project uses the Go standard logger. It's certainly debatable if logs are "conventional output for a server application", since it's not comparable with say a CLI application. There has been a nice conversation on reddit about exactly this topic: https://www.reddit.com/r/golang/comments/15ndgdy/slog_splitting_stdoutstderr_streams/

Since Go 1.21, structured logging comes right out of the box ([log/slog](https://pkg.go.dev/log/slog)). I haven't had the time yet to properly check out slog, but I assume that we should be using it. While we're at that, we could also change the behavior so that regular logs go to stdout and warnings/errors go to stderr.

To answer your question:
> Is it expected behavior to log this into stderr instead of stdout ?

It is expected, but not necessarily intended in the long term.

**Comment by uggyuggy** [comment]
Thank you for the detailed answer.
